lib/Dialect/Poly/IR/PolyOps.cpp

```diff
@@ -0,0 +1,29 @@
+#include "include/Dialect/Poly/IR/PolyOps.h"
+
+namespace mlir {
+namespace heir {
+namespace poly {
+
+void ExtractSliceOp::build(OpBuilder &builder, OperationState &result,
+                           Value input, Value i, Value j) {
+  PolynomialType polyType = llvm::dyn_cast<PolynomialType>(input.getType());
+  RingAttr attr = polyType.getRing();
+  uint32_t idealDegree = attr.ideal().getDegree();
+  IntegerType elementTy = builder.getIntegerType(
+      attr.coefficientModulus().getBitWidth(), /*isSigned=*/false);
+  Type resultType = RankedTensorType::get({idealDegree}, elementTy, attr);
+  build(builder, result, resultType, input, i, j);
+}
+
+void PolyFromCoeffsOp::build(OpBuilder &builder, OperationState &result,
+                             Value input, RingAttr ring) {
+  TensorType tensorType = dyn_cast<TensorType>(input.getType());
+  APInt cmod(APINT_BIT_WIDTH, 1);
+  cmod = cmod << tensorType.getElementTypeBitWidth();
+  Type resultType = PolynomialType::get(builder.getContext(), ring);
+  build(builder, result, resultType, input);
+}
```

Comment:
What about using InferTypeOpInterface instead? That way, you only have to implement the actual type inference, with the builder(s) still being auto generated? See https://github.com/MarbleHE/HECO/blob/main/include/heco/IR/FHE/FHE.td#L135-L138

Reply:
Just learning about this now. Will add it.

Reply:
On closer inspection, PolyFromCoeffsOp needs the special builder because the ring cannot be inferred from the tensor alone. I migrated ExtractSliceOp, though.